src/plugins/discover/public/application/main/components/top_nav/get_top_nav_links.tsx

```diff
@@ -159,17 +158,42 @@ export const getTopNavLinks = ({
     }),
     testId: 'shareTopNavButton',
     run: async (anchorElement: HTMLElement) => {
-      const updatedDataView = await persistDataView(dataView);
-      if (!services.share || !updatedDataView) {
-        return;
-      }
-      const sharingData = await getSharingData(searchSource, state.appState.getState(), services);
+      if (!services.share) return;
 
+      const { locator } = services;
+      const appState = state.appState.getState();
+      const { timefilter } = services.data.query.timefilter;
+      const timeRange = timefilter.getTime();
+      const refreshInterval = timefilter.getRefreshInterval();
+      const { grid, ...otherState } = appState;
+
+      // Share -> Get links -> Snapshot
+      const params: DiscoverAppLocatorParams = {
+        ...otherState,
+        ...(savedSearch.id ? { savedSearchId: savedSearch.id } : {}),
+        ...(dataView?.isPersisted()
+          ? { dataViewId: dataView?.id }
+          : { dataViewSpec: dataView?.toSpec() }),
+        timeRange,
+        refreshInterval,
+      };
+      const relativeUrl = locator.getRedirectUrl(params);
+      const shareableUrl = relativeToAbsolute(relativeUrl);
+
+      // Share -> Get links -> Saved object
+      const shareableUrlForSavedObject = await locator.getUrl(
+        { savedSearchId: savedSearch.id },
+        { absolute: true }
+      );
+
+      const sharingData = await getSharingData(searchSource, appState, services);
       services.share.toggleShareContextMenu({
         anchorElement,
         allowEmbed: false,
         allowShortUrl: !!services.capabilities.discover.createShortUrl,
-        shareableUrl: unhashUrl(window.location.href),
+        shareableUrl,
+        shareableUrlForSavedObject,
+        shareableUrlLocatorParams: { locator, params },
```

Comment:
Having this `shareableUrlLocatorParams` in place, couldn't part of the logic to actually get the URL be migrated into the sharing service?

Reply:
https://github.com/elastic/kibana/pull/154947/commits/985e522f1086cd13153f7ca382e5bb8a4c30c31b

@vadimkibana Do you mind taking a look at those changes and see if they are acceptable?

Reply:
I've reverted this change. It seemed to introduce some other bugs, and I'd prefer to keep this PR simpler. I can create a follow-up issue when this is merged.

Reply:
makes sense 👍